pkg/k8s/endpoints.go

```
@@ -239,6 +241,54 @@ func parseEndpointPort(port slim_discover_v1beta1.EndpointPort) (string, *loadba
 	return name, lbPort
 }
 
+// endpointSlices is the collection of all endpoint slices in the cluster.
+// The map key is the name of the endpoint slice.
+type endpointSlices struct {
+	epSlices map[string]*Endpoints
+}
+
+// newEndpointsSlices returns a new endpointSlices
+func newEndpointsSlices() *endpointSlices {
+	return &endpointSlices{
+		epSlices: map[string]*Endpoints{},
+	}
+}
+
+// GetEndpoints returns a read only a single *Endpoints structure with all
+// Endpoints' backends joined.
+func (es *endpointSlices) GetEndpoints() *Endpoints {
+	if es == nil || len(es.epSlices) == 0 {
+		return nil
+	}
+	allEps := newEndpoints()
+	for _, eps := range es.epSlices {
+		for backend, ep := range eps.Backends {
+			allEps.Backends[backend] = ep
+		}
+	}
+	return allEps
+}
+
+// Upsert maps the 'esname' to 'e'.
+// - 'esname': Name of the Endpoint Slice
+// - 'e': Endpoints to store in the map
+func (es *endpointSlices) Upsert(esname string, e *Endpoints) {
+	if es == nil {
+		panic("nil map")
```

Do we really want to panic here? Maybe return an error instead, so that the top layer can decide what to do?

---

IMO, yes, if this happens it is a programmer error.

---

Then maybe we could change the panic msg to "BUG: endpointSlices is nil"?